samples/cpp/flann_search_dataset.cpp

```diff
@@ -0,0 +1,264 @@
+                          static_cast<float>(dists.at<DIST_TYPE>(i,0)));
+            good_matches.push_back( dmatch );
+            matches_per_img_histogram[ db_indice_2_image_lut[indice_in_db] ]++;
+        }
+    }
+
+
+    //-- Step 4: find the dataset image with the highest proportion of matches
+    std::multimap<float, img_info> images_infos;
+    for (int i = 0; i < nbr_of_imgs; ++i)
+    {
+        const unsigned int nbr_of_matches = matches_per_img_histogram[i];
+        if (nbr_of_matches < 4) //we need at leat 4 points for a homography
+            continue;
+
+        const unsigned int nbr_of_kpts = db_images_indice_range[i+1] - db_images_indice_range[i];
+        const float inverse_proportion_of_retrieved_kpts =
+                static_cast<float>(nbr_of_kpts) / static_cast<float>(nbr_of_matches);
+
+        img_info info(i, nbr_of_matches);
+        images_infos.insert( std::pair<float,img_info>(inverse_proportion_of_retrieved_kpts,
+                                                       info) );
+    }
+
+    if (images_infos.begin() == images_infos.end())
+    {
+        cout<<"No good match could be found."<<endl;
+        return 0;
+    }
+
+    //-- if there are several images with a similar proportion of matches,
+    // select the one with the highest number of matches weighted by the
+    // squared ratio of proportions
+    const float best_matches_proportion = images_infos.begin()->first;
+    float new_matches_proportion = best_matches_proportion;
+    img_info best_img = images_infos.begin()->second;
+
+    std::multimap<float, img_info>::iterator it = images_infos.begin();
+    ++it;
+    while ((it->first < 1.1*best_matches_proportion) && (it!=images_infos.end()))
+    {
+        const float ratio = new_matches_proportion / it->first;
+        if( it->second.nbr_of_matches * (ratio * ratio) > best_img.nbr_of_matches)
+        {
+            new_matches_proportion = it->first;
+            best_img = it->second;
+        }
+        ++it;
+    }
+
+    //-- Step 5: filter goodmatches that belong to the best image match of the dataset
+    std::vector<DMatch> filtered_good_matches;
+    for (std::vector<DMatch>::iterator itr(good_matches.begin()); itr != good_matches.end(); ++itr)
+    {
+        if (itr->imgIdx == best_img.img_index)
+            filtered_good_matches.push_back(*itr);
+    }
+
+    //-- Retrieve the best image match from the dataset
+    Mat db_img = imread( files[best_img.img_index], IMREAD_GRAYSCALE );
+
+    //-- Draw matches
+    Mat img_matches;
+    drawMatches( img, img_keypoints, db_img, db_keypoints, filtered_good_matches, img_matches, Scalar::all(-1),
+                 Scalar::all(-1), std::vector<char>(), DrawMatchesFlags::NOT_DRAW_SINGLE_POINTS );
+
+    //-- Show detected matches
+    imshow("Good Matches", img_matches );
+
+    while(1)
+    {
+        int k = waitKey();
+        if (k == 27) // Esc key
+            break;
+    }
+    delete index;
```

Comment:
> delete

Lets avoid using of manual memory management.

Please use `std::shared_ptr` approach instead (through `cv::Ptr` / `makePtr` emulation for C++98).